contracts/p1/RevenueTrader.sol

```diff
@@ -89,14 +119,121 @@ contract RevenueTraderP1 is TradingP1, IRevenueTrader {
         );
 
         if (launch) {
-            tryTrade(req);
+            openTrade(req);
+        }
+    }
+
+    /// Executes an atomic swap for revenue via a dutch auction
+    /// @dev Caller must have granted tokenIn allowances for required tokenIn bal
+    /// @param tokenIn The ERC20 token provided by the caller
+    /// @param tokenOut The ERC20 token being purchased by the caller
+    /// @param amountOut {qTokenOut} The exact quantity of tokenOut being purchased
+    /// @return s The exact Swap performed
+    /// @custom:interaction RCEI
+    function swap(
+        IERC20 tokenIn,
+        IERC20 tokenOut,
+        uint256 amountOut
+    ) external notTradingPausedOrFrozen returns (Swap memory s) {
+        // == Refresh ==
+        assetRegistry.refresh();
+        furnace.melt();
+
+        require(address(trades[tokenOut]) == address(0), "nonatomic trade ongoing");
+        require(tokenIn == tokenToBuy, "will only buy tokenToBuy");
+        require(tokenOut != tokenToBuy, "will not sell tokenToBuy");
+
+        // executeSwap if storage auction already exists
+        DutchAuction storage auction = dutchAuctions[tokenOut][tradeEnd];
+        if (
+            tradeEnd > block.timestamp &&
+            (address(auction.sell) != address(0) || address(auction.buy) != address(0))
```

Comment:
do we even need `address(auction.buy) != address(0)`? don't we just need to make sure that the sell token is correct?

Reply:
In P1 I'm going to have it just do `address(auction.sell) != address(0)` since it will be an external call to compare `auction.sell.erc20()` against tokenOut, but in P0 I'll do it the other way.